extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/YoutubeParsingHelper.java

```diff
@@ -393,68 +431,87 @@ private static void extractClientVersionAndKey() throws IOException, ExtractionE
             }
         }
 
-        String contextClientVersion;
-        final String[] patterns = {
-                "INNERTUBE_CONTEXT_CLIENT_VERSION\":\"([0-9\\.]+?)\"",
-                "innertube_context_client_version\":\"([0-9\\.]+?)\"",
-                "client.version=([0-9\\.]+)"
-        };
-        for (final String pattern : patterns) {
-            try {
-                contextClientVersion = Parser.matchGroup1(pattern, html);
-                if (!isNullOrEmpty(contextClientVersion)) {
-                    clientVersion = contextClientVersion;
-                    break;
-                }
-            } catch (final Parser.RegexException ignored) {
-            }
+        try {
+            clientVersion = getStringResultFromRegexArray(html,
+                    INNERTUBE_CONTEXT_CLIENT_VERSION_REGEXES, 1);
+        } catch (final Parser.RegexException ignored) {
         }
 
         if (!isNullOrEmpty(clientVersion) && !isNullOrEmpty(shortClientVersion)) {
             clientVersion = shortClientVersion;
         }
 
         try {
-            key = Parser.matchGroup1("INNERTUBE_API_KEY\":\"([0-9a-zA-Z_-]+?)\"", html);
-        } catch (final Parser.RegexException e1) {
-            try {
-                key = Parser.matchGroup1("innertubeApiKey\":\"([0-9a-zA-Z_-]+?)\"", html);
-            } catch (final Parser.RegexException e2) {
-                throw new ParsingException("Could not extract client version and key");
-            }
+            key = getStringResultFromRegexArray(html, INNERTUBE_API_KEY_REGEXES, 1);
+        } catch (final Parser.RegexException e) {
+            throw new ParsingException(
+                    "Could not extract YouTube WEB InnerTube client version and API key from HTML search results page");
         }
         keyAndVersionExtracted = true;
     }
 
     /**
-     * Get the client version
+     * Get the client version used by YouTube website on InnerTube requests.
+     *
+     * @throws IOException         if something goes wrong when getting the client version
+     * @throws ExtractionException if something goes wrong when getting the client version
      */
     public static String getClientVersion() throws IOException, ExtractionException {
-        if (!isNullOrEmpty(clientVersion)) return clientVersion;
-        if (areHardcodedClientVersionAndKeyValid()) {
-            return clientVersion = HARDCODED_CLIENT_VERSION;
+        if (!isNullOrEmpty(clientVersion)) {
+            return clientVersion;
+        }
+
+        try {
+            extractClientVersionAndKeyFromSwJs();
+        } catch (final Exception e) {
+            extractClientVersionAndKeyFromHtmlSearchResultsPage();
         }
 
-        extractClientVersionAndKey();
-        return clientVersion;
+        if (keyAndVersionExtracted) {
+            return clientVersion;
+        } else {
+            if (areHardcodedClientVersionAndKeyValid()) {
```

Comment:
To prevent YouTube fingerprinting on videoplayback URLs based on the client version used to request the `player` endpoint (it seems the use of an outdated client version was one of the origins of buffering issues with streams of the `WEB` client).